### PR TITLE
Added some instructions on using MinGW-built XGBoost with python.

### PR DESCRIPTION
**Pull request by KOLANICH**

### CODE REVIEW

File: doc/build.rst [hunk 1]
```diff
@@ -161,6 +161,21 @@ To build with MinGW, type:
 
   cp make/mingw64.mk config.mk; make -j4
 
+Opening a can of worms - building XGBoost library for python for Windows with MinGW-w64
+---------------------------------------------------------------------------------------
+
+Windows versions of python are built with Microsoft proprietary Visual Studio. This has multiple issues:
+
+1. VS is proprietary and commercial software (it may be illegal to install the free version in your org). It is bundled with telemetry. It also has been spotted on inserting telemetry into apps.
```
Comment (hcho3) [inline doc/build.rst]: Let's re-phrase this bullet point as follows:

1. VS is proprietary and commercial software. While Microsoft provides a free edition (Community Edition), installing it may be against the policy of your organization. In addition, Visual Studio contains telemetry, as documented in Microsoft Visual Studio Licensing Terms (https://visualstudio.microsoft.com/license-terms/mt736442/). This has raised privacy and security concerns among users and system administrators.

I took out the part about telemetry insertion into apps, since Microsoft has provided a response to the allegation: https://www.reddit.com/r/cpp/comments/4ibauu/visual_studio_adding_telemetry_function_calls_to/.

File: doc/build.rst [hunk 2]
```diff
@@ -161,6 +161,21 @@ To build with MinGW, type:
...
+2. g++ usually generates faster code on `-O3`.
+
+So you may want to build xgboost with g++ own yohr own risk. This opens a can of worms because MSVC uses Microsoft runtime and MinGW-w64 uses own runtime, and runtimes have differrent memory allocators. But in fact this setup is usable if you know how to deal with it. Here is some experience.
```
Comment (hcho3) [inline doc/build.rst]: Typo: yohr -> your, differrent -> different

File: doc/build.rst [hunk 3]
```diff
@@ -164,16 +164,18 @@ To build with MinGW, type:
...
 ---------------------------------------------------------------------------------------
 
-Windows versions of python are built with Microsoft proprietary Visual Studio. This has multiple issues:
+Windows versions of python are built with Microsoft Visual Studio. This has multiple issues:
 
-1. VS is proprietary and commercial software (it may be illegal to install the free version in your org). It is bundled with telemetry. It also has been spotted on inserting telemetry into apps.
+1. VS is proprietary and commercial software (it may be illegal to install the free version in your org). While Microsoft provides a free edition (Community Edition), installing it may be illegal and/or against the policy of your organization. It is bundled with telemetry. Visual Studio contains telemetry, as documented in [Microsoft Visual Studio Licensing Terms] (https://visualstudio.microsoft.com/license-terms/mt736442/). It also [has been spotted on inserting telemetry](https://old.reddit.com/r/cpp/comments/4ibauu/visual_studio_adding_telemetry_function_calls_to/) into apps. These facts have raised privacy and security concerns among some users and system administrators.
```
Comment (hcho3) [inline doc/build.rst]: > (it may be illegal to install the free version in your org)

This is redundant, because of the next sentence

File: doc/build.rst [hunk 4]
```diff
@@ -161,6 +161,24 @@ To build with MinGW, type:
 
   cp make/mingw64.mk config.mk; make -j4
 
+Opening a can of worms - building XGBoost library for python for Windows with MinGW-w64
+---------------------------------------------------------------------------------------
+
+Windows versions of python are built with Microsoft Visual Studio. This has multiple issues:
+
+1. VS is proprietary and commercial software. While Microsoft provides a free edition (Community Edition), installing it may be illegal in your organization.
+2. Visual Studio contains telemetry, as documented in `Microsoft Visual Studio Licensing Terms <https://visualstudio.microsoft.com/license-terms/mt736442/>`_. It `has been spotted on inserting telemetry <https://old.reddit.com/r/cpp/comments/4ibauu/visual_studio_adding_telemetry_function_calls_to/>`_ into apps. At least the Community Edition cannot even be easily downloaded without running the app collecting telemetry. These facts have raised privacy and security concerns among some users and system administrators. Running this kind of software may be against the policy of your organization.
```
Comment (hcho3) [inline doc/build.rst]: > At least the Community Edition cannot even be easily downloaded without running the app collecting telemetry

This is redundant, since we stated earlier that VS contains telemetry.

> It `has been spotted on inserting telemetry <https://old.reddit.com/r/cpp/comments/4ibauu/visual_studio_adding_telemetry_function_calls_to/>`_ into apps.

Are we keeping this line?

Comment (hcho3) [reply doc/build.rst]: > It has been spotted on inserting telemetry <https://old.reddit.com/r/cpp/comments/4ibauu/visual_studio_adding_telemetry_function_calls_to/>_ into apps.

Let's rephrase this as "Visual Studio has also inserted telemetry into apps for the purpose of instrumentation."

> Running this kind of software may be against the policy of your organization.

Rephrase this as: It may be against the policy of your organization to run software containing telemetry.

In general, we do not want to make strong value judgment on this page. Let's use neutral tone, objectively describe the existence of telemetry, and let users decide whether it's a deal breaker or not.

File: doc/build.rst [hunk 5]
```diff
@@ -161,6 +161,24 @@ To build with MinGW, type:
 
   cp make/mingw64.mk config.mk; make -j4
 
+Opening a can of worms - building XGBoost library for python for Windows with MinGW-w64
```
Comment (hcho3) [inline doc/build.rst]: Can we remove the phrase "Opening a can of worms" from the title, to keep it short?

File: doc/build.rst [hunk 6]
```diff
@@ -161,6 +161,24 @@ To build with MinGW, type:
...
+---------------------------------------------------------------------------------------
+
+Windows versions of python are built with Microsoft Visual Studio. This has multiple issues:
```
Comment (hcho3) [inline doc/build.rst]: "Windows versions of python are built with Microsoft Visual Studio, raising several potential concerns."

File: doc/build.rst [hunk 7]
```diff
@@ -161,6 +161,24 @@ To build with MinGW, type:
...
+
+1. VS is proprietary and commercial software. While Microsoft provides a free edition (Community Edition), installing it may be illegal in your organization.
```
Comment (hcho3) [inline doc/build.rst]: Change this to  "VS is proprietary and commercial software. While Microsoft provides a free edition (Community Edition), your organization may prohibit its use because of its licensing terms."

File: doc/build.rst (hunk 4 above)
Comment (KOLANICH) [reply doc/build.rst]: >This is redundant, since we stated earlier that VS contains telemetry.

Not quite.

In order to download VS one has to run an app with telemetry. They provide no offline installer. Instead one runs their app which downloads the msi packages (and there is no easy control on what it downloads and installs, for example I don't need arm toolchain, but it still downloads it) and sends telemetry. So even if one wants to put VS into a sandbox/VM isolated from inet (whkch should prevent leaking telemetry), he will have to create a VM with inet and expect installer in it to send telemetry.

Comment (hcho3) [reply doc/build.rst]: Okay, how about "the Visual Studio installer also contains telemetry, requiring internet connection for installation."

File: doc/build.rst (hunk 7 above)
Comment (KOLANICH) [reply doc/build.rst]: The problem is not in the organization. An org may have no policy about software at all.

Comment (hcho3) [reply doc/build.rst]: Did you mean that installing VS on a company computer may be against the licensing terms?

Comment (hcho3) [reply doc/build.rst]: FYI, according to https://visualstudio.microsoft.com/license-terms/mlt553321/,

> If you are an enterprise, your employees and contractors may not use the software to develop or test your applications, except for: (i) open source; (ii) Visual Studio extensions; (iii) device drivers for the Windows operating system; and, (iv) education purposes as permitted above

This is quite stringent restriction.

Comment (hcho3) [reply doc/build.rst]: If the MS licensing terms are problem, we can state that "While Microsoft provides a free edition (Community Edition), their licensing terms restrict when, where, and how it can be installed and used."

**Comment by hcho3** [comment]
@KOLANICH Your latest fix looks good. I'll merge it once tests pass. Thanks!
